flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/sink/ConstraintValidator.java

```diff
@@ -0,0 +1,209 @@
+            checkArgument(
+                    notNullFieldIndices.length > 0,
+                    "ConstraintValidator requires that there are not-null fields.");
+            this.notNullFieldIndices = notNullFieldIndices;
+            this.notNullEnforcer = notNullEnforcer;
+            this.allFieldNames = allFieldNames;
+            if (notNullEnforcer != null) {
+                this.mustApply = true;
+            }
+        }
+
+        public void addCharPrecisionConstraint(
+                CharPrecisionEnforcer charPrecisionEnforcer,
+                List<Tuple2<Integer, Integer>> charFields,
+                String[] allFieldNames) {
+            this.charPrecisionEnforcer = charPrecisionEnforcer;
+            if (this.charPrecisionEnforcer == CharPrecisionEnforcer.TRIM) {
+                checkArgument(
+                        charFields.size() > 0,
+                        "ConstraintValidator requires that there are CHAR/VARCHAR fields.");
+                this.charFields = charFields;
+                this.allFieldNames = allFieldNames;
+                this.mustApply = true;
+            }
+        }
+
+        public ConstraintValidator build() {
+            return new ConstraintValidator(
+                    notNullEnforcer,
+                    notNullFieldIndices,
+                    charPrecisionEnforcer,
+                    charFields != null ? charFields.stream().mapToInt(t -> t.f0).toArray() : null,
+                    charFields != null ? charFields.stream().mapToInt(t -> t.f1).toArray() : null,
+                    allFieldNames);
+        }
+    }
+
+    @Override
+    public void processElement(StreamRecord<RowData> element) throws Exception {
+        StreamRecord<RowData> processedElement = processNotNullConstraint(element);
+        if (processedElement != null) {
+            output.collect(processCharConstraint(processedElement));
+        }
+    }
+
+    private StreamRecord<RowData> processNotNullConstraint(StreamRecord<RowData> element) {
+        if (notNullEnforcer == null) {
+            return element;
+        }
+
+        final RowData rowData = element.getValue();
+
+        for (int index : notNullFieldIndices) {
+            if (rowData.isNullAt(index)) {
+                switch (notNullEnforcer) {
+                    case ERROR:
+                        throw new TableException(
+                                String.format(
+                                        "Column '%s' is NOT NULL, however, a null value is being written into it. "
+                                                + "You can set job configuration '%s'='drop' "
+                                                + "to suppress this exception and drop such records silently.",
+                                        allFieldNames[index],
+                                        TABLE_EXEC_SINK_NOT_NULL_ENFORCER.key()));
+                    case DROP:
+                        return null;
```

Comment:
Dropping a record without logging to me it sounds like an open StackOverflow question already :smile:. Perhaps what about adding another `LOG_AND_DROP` strategy to `NotNullEnforcer`? In case user uses `DROP`, then we log nothing, otherwise with `LOG_AND_DROP` we log the broken record

Reply:
That's the existing behaviour though, so if we want to change/improve, created: https://issues.apache.org/jira/browse/FLINK-25008 for future improvement.